<!--
%\VignetteEngine{knitr}
%\VignetteIndexEntry{An Introduction to the multiseq package}
-->

An Introduction to **multiseq** package
=======================================


```{r setup, include=FALSE}
library(knitr)
opts_chunk$set(out.extra='style="display:block; margin: auto"', fig.align="center")
```


Introduction
------------


The  **multiseq** package is an **R** package for multiscale sequence analysis. Multiseq has two main modes of operation: smoothing and ,...... Given one (or more) observed one dimensional signal(s) generated by an underlying Poisson distribution multiseq can provide an estimate of the mean and standard error of the original distribution. Given two groups of observed one dimensional signals generated by two underlying Poisson distributions multiseq can provide an estimate of the mean difference and standard error between the two distributions. The main assumption of ... is that ... graphical display of a correlation matrix, confidence interval. 

The  **multiseq** package is ongoing work in the [Stephens lab](http://stephenslab.uchicago.edu) at the University of Chicago.

To download the package go to [link](https://github.com/stephenslab/multiseq/tree/master/package/multiseq.tar.gz) and click on "View Raw" or click on this [link](https://github.com/stephenslab/multiseq/blob/master/package/multiseq.tar.gz?raw=true).

After uncompressing the downloaded archive into folder `multiseq`, you can find installation instructions in `multiseq/README.md`. 

Once you have successfully installed the package following instructions, you can load the package with:

```{r loading}
    library(multiseq)
```

Loading data
------------

```{r load_data}
    #load example data - R object
    data(OAS1,package="multiseq")
    x <- OAS1$M
    g <- OAS1$g
    read.depth <- OAS1$read.depth
```

Loading sequencing data using function `get.counts` 
---------------------------------------------------
**Multiseq** provides the user with a function to load sequencing data in `bam`, `bigWig` or `hdf5` format. The function is called `get.counts` and requires as input a sample sheet and a genomic region.

The samplesheet should have the following format (see file ............data/sim/samplesheet.sim.txt).........bed peaks wig see trackhub.R:

    SampleID Type Tissue Replicate Peaks ReadDepth bigWigPath
    ......055A RNASeq ......24hrShamControl 8 - 1550735 ........data/sim/055A.bw
    ......055B RNASeq .......24hr2uMsimvastatinLPDS 8 - 2350343 ........data/sim/055BNonNull.bw
    .......056A RNASeq ........24hrShamControl 9 - 1320166 .......data/sim/056A.bw
    ......056B RNASeq ..........24hr2uMsimvastatinLPDS 9 - 1723647 ...........data/sim/056BNonNull.bw

```{r getcounts}
    samplesheet <- "........data/sim/samplesheet.sim.txt"
    region      <- "chr5:131989505-132120576"

    samples     <- read.table(samplesheet, stringsAsFactors=F, header=T) 
    g           <- factor(samples$Tissue) 
    g           <- match(g, levels(g))-1
    M           <- get.counts(samples, region) 
```

Smoothing using function `multiseq`
-----------------------------------

In this section we will show how **multiseq** can be used to smooth a one dimensional signal:

```{r testing_smoothing}
    #run multiseq on data
    res0 <- multiseq(x=x, g=g, minobs=1, lm.approx=FALSE, read.depth=read.depth)
    
    #plot estimated mean value +- 2 s.d.
    fra <- 2 #s.d.
    plotResults(res0, fra) #?....... type="baseline")
```

Finding differential signal using function `multiseq`
-----------------------------------------------------

```{r testing_diff}
    res <- multiseq(x=x, g=g, minobs=1, lm.approx=FALSE, read.depth=read.depth)

    #plot estimated effect and s.d. and print intervals where there is an effect
    plotResults(res, fra)
    get.effect.intervals(res,fra)

    #save results in dir.name
    dir.name="~/src/multiseq/data/multiseq_sim/"


    # this function saves results in file effect_mean_var.txt.gz, a file with two columns: first column is effect mean and second column is effect variance
    write.effect.mean.variance.gz(res, dir.name)
    # write a bed file with intervals where multiseq found an effect at 2 sd 
    write.effect.intervals(res, dir.name, fra)


```

Visualizing input data in the UCSC Genome Browser
-------------------------------------------------

With function `samplesheetToTrackHub` you can create a (Track Hub)[http://www.genome.ucsc.edu/goldenPath/help/hgTrackHubHelp.html] and visualize the input data in the UCSC Genome Browser.

```{r track_hub}
    hub_name < -"testMultiseq/sim"
    samplesheetToTrackHub(samplesheet,hub_name)
``

This will create a Track Hub in `/some/path/testMultiseq/sim/` and will print the following message:

    go to http://genome.ucsc.edu/cgi-bin/hgHubConnect and click on the My Hubs window    
    copy paste the following string in the URL field
    https:some/address/testMultiseq/sim/hub.txt
    center the genome browser on the region of interest
    if the track is hidden click on show and then refresh

If the read tracks or the bed files are large, make sure enough memory is available to run `simulationToTrackHub` (e.g., use ql 10g on the PPS cluster).

This is a screenshot of the data in the Genome Browser:
![Image](../../../data/sim/sim.png?raw=true)

Visualizing output data in the UCSC Genome Browser
-------------------------------------------------

After running multiseq and saving results using `write.effect.mean.variance.gz` and `write.effect.intervals` (see above), you can create a Track Hub of results using function `multiseqToTrackHub`. `multiseqToTrackHub` will create a Track Hub that can be visualized in the UCSC Genome Browser to display
- the effect +- 2 standard errors
- the significant intervals at 2 sd.

In order to run function `multiseqToTrackHub` on multiseq output you need to specify a few arguments: the multiseq output folder multiseq_folder, the path to a file with chromosome names and lengths chrom_file

```{r multiseqToTrackHub}
    region          <- "chr5:131989505-132120576"
    hub_name        <- "testMultiseq/multiseq_sim"
    multiseq_folder <- "~/src/multiseq/data/multiseq_sim"
    chrom_file      <- "~/src/multiseq/data/chromosome.lengths.hg19.txt"
    multiseqToTrackHub(region, hub_name, multiseq_folder, chrom_file)
```

Function `multiseqToTrackHub` will create a Track Hub named *multiseq_sim* in the `https:some/address/testMultiseq/` folder and will print the following message:
  
    go to http://genome.ucsc.edu/cgi-bin/hgHubConnect and click on the My Hubs window
    copy paste the following string in the URL field
    https:some/address/testMultiseq/multiseq_sim/hub.txt
    note: center your genome browser around chr5:131989505-132120576 and make track visible

This is a screenshot of the Track Hub from the Genome Browser:
![Image](data/sim/multiseq.png?raw=true)

$$latex
a_i = 
\begin{cases}
			\tan (e_{i2}/e_{i1}), & \text{if $e_{i1}>0$;}
			     		       \newline
							\tan (e_{i2}/e_{i1}) + \pi, & \text{otherwise.}
\end{cases}						     		     
$$ 

where $e_1$ and $e_2$ are the largest two eigenvalues of the correlation  matrix. See [Michael Friendly (2002)](www.datavis.ca/papers/corrgram.pdf) for details.

* `"FPC"` for the first principal component order.

* `"hclust"` for hierarchical clustering order, and `"hclust.method"` for the agglomeration method to be used . `"hclust.method"` should be one of `"ward"`, `"single"`, `"complete"`, `"average"`, `"mcquitty"`, `"median"` or `"centroid"`.

* `"alphabet"` for alphabetical order.

```{r order}
corrplot(M, order ="AOE")
corrplot(M, order ="hclust")
corrplot(M, order ="FPC")
corrplot(M, order ="alphabet")
```

If using `"hclust"`, `corrplot()` can  draw rectangles around the chart of corrrlation matrix based on the results of  hierarchical clustering.

```{r rectangles}
corrplot(M, order="hclust", addrect=2)
corrplot(M, order="hclust", addrect=3)
```
!!!

Here is an animation to show the relation between significant level and confidence interval.
```{r ani, eval=FALSE}
for(i in seq(0.1, 0, -0.005)){
  tmp <- cor.mtest(mtcars,1-i)
  corrplot(M, p.mat = tmp[[1]], low=tmp[[2]], upp=tmp[[3]], order="hclust",
  	      pch.col="red", sig.level = i, plotC="rect", cl.pos="n",
	      		     mar=c(0,0,1,0), 
			     		     title=substitute(alpha == x,list(x=format(i,digits=3,nsmall=3))))
}
```

<div align = "center">
 <embed width="504" height="504" name="plugin" src="http://animation.r-forge.r-project.org/swf/corrplot-ani.swf" type="application/x-shockwave-flash"> 
</div>

