\name{samplesheetToTrackHub}
\alias{samplesheetToTrackHub}
\title{Create a UCSC genome browser "Track Hub" from read tracks and bed tracks listed in a samplesheet.

This function requires the executables "wigToBigWig", "bedToBigBed" and "bigWigInfo" to be in the user's PATH.
Read tracks can be in bam, hdf5, wig or bigwig format and significant intervals can be in bed format.}
\usage{
samplesheetToTrackHub(samplesheet, hub_name = NULL,
  chrom_file = "~/src/multiseq/data/chromosome.lengths.hg19.txt",
  assembly = "hg19", mountpoint = MOUNTPOINT_PATH,
  http_address = MOUNTPOINT_HTTP_ADDRESS)
}
\arguments{
  \item{samplesheet:}{path to the samplesheet; the
  samplesheet must contain a column with header sampleID
  and: either a column with header h5FullPath, or a column
  with header bigWigPath containing the path to the hdf5
  files or the bigWig files, respectively. If the
  samplesheet has a column with header Peaks it must also
  have a column with header Tissue. Depending on the size
  of the hdf5 files this code might require a lot of
  memory}

  \item{hub_name:}{name of the track hub; it could contain
  a path, in which case the path will be relative to the
  mountpoint(see below). This string can be set to any
  value; defauly=paste0(basename(samplesheet),".TrackHub")}

  \item{chrom_file:}{path to the file containing chromosome
  names and lengths;
  default="~/src/multiseq/data/chromosome.lengths.hg19.txt"}

  \item{assembly:}{genome assembly that reads were mapped
  to; default="hg19"}

  \item{mountpoint:}{path to the directory where the track
  hub folder will be saved in. This directory should be
  associated with an http address or an ftp address;
  default=Sys.getenv("MOUNTPOINT_PATH")}

  \item{http_address:}{http or ftp address associated with
  the mountpoint; default=Sys.getenv("MOUNTPOINT_PATH")}
}
\description{
Create a UCSC genome browser "Track Hub" from read tracks
and bed tracks listed in a samplesheet.

This function requires the executables "wigToBigWig",
"bedToBigBed" and "bigWigInfo" to be in the user's PATH.
Read tracks can be in bam, hdf5, wig or bigwig format and
significant intervals can be in bed format.
}

